crates/board/src/crypto.rs

```diff
@@ -158,8 +177,79 @@ impl crate::Supported for sha2::Sha256 {}
 impl crate::Supported for sha2::Sha384 {}
 
 #[cfg(feature = "internal-software-crypto-hmac")]
-impl<D: Support<bool> + Default + BlockSizeUser + Update + FixedOutput + HashMarker> Support<bool>
-    for hmac::SimpleHmac<D>
+impl<D: Support<bool> + Default + BlockSizeUser + Update + FixedOutputReset + HashMarker>
+    Support<bool> for hmac::SimpleHmac<D>
 {
     const SUPPORT: bool = D::SUPPORT;
 }
+
+/// Extends APIs with infallible operations to support errors.
+///
+/// Infallible operations may return incorrect values. Those results must not be used until
+/// `last_error()` is called and returns `Ok(())`, signifying that the return value is correct.
+#[cfg(any(feature = "internal-api-crypto-hash", feature = "internal-api-crypto-hmac"))]
+pub trait LastError {
+    /// Returns the last error of infallible operations.
+    fn last_error(&self) -> Result<(), Error>;
+}
+
+/// Wrapper API around Hash that calls last_error.
+#[cfg(feature = "internal-api-crypto-hash")]
+pub struct HashApi<T: Hash>(pub T);
+
+/// Wrapper API around Hmac that calls last_error.
+#[cfg(feature = "internal-api-crypto-hmac")]
+pub struct HmacApi<T: Hmac>(pub T);
+
+#[cfg(feature = "internal-api-crypto-hash")]
+impl<T: Hash> OutputSizeUser for HashApi<T> {
+    type OutputSize = T::OutputSize;
+}
+
+#[cfg(feature = "internal-api-crypto-hmac")]
+impl<T: Hmac> OutputSizeUser for HmacApi<T> {
+    type OutputSize = T::OutputSize;
+}
+
+#[cfg(feature = "internal-api-crypto-hash")]
+impl<T: Hash> HashApi<T> {
+    /// Constructor of HmacApi.
+    pub fn new() -> Result<Self, Error> {
+        let hash = T::default();
+        hash.last_error().map(|()| HashApi(hash))
+    }
+
+    /// Call update from RustCrypto API and then last_error.
```

Comment:
nit: I think the type documentation is enough context regarding last_error, the function type also reminds it. We can just document the high-level operation.
```suggestion
    /// Updates the hash with the provided data.
```